src/nvim/lua/vim.lua

```diff
@@ -118,11 +118,84 @@ local function _update_package_paths()
   last_nvim_paths = cur_nvim_paths
 end
 
+local function gsplit(s, sep, plain)
+  assert(type(s) == "string")
+  assert(type(sep) == "string")
+  assert(type(plain) == "boolean" or type(plain) == "nil")
+
+  local start = 1
+  local done = false
+
+  local function pass(i, j, ...)
+    if i then
+      assert(j+1 > start, "Infinite loop detected")
+      local seg = s:sub(start, i - 1)
+      start = j + 1
+      return seg, ...
+    else
+      done = true
+      return s:sub(start)
+    end
+  end
+
+  return function()
+    if done then
+      return
+    end
+    if sep == '' then
+      if start == #s then
+        done = true
+      end
+      return pass(start+1, start)
+    end
+    return pass(s:find(sep, start, plain))
+  end
+end
+
+local function split(s,sep,plain)
```

Comment:
Maybe the parameter should be named `sep_pat`. Pattern behavior can be quite surprising when one was expecting a string, so having a convention that signals to users "this has pattern behavior" could help a lot.